### ISSUE TITLE
wxOSX Cocoa: Running application not in a .app bundle does not gain focus on startup

### ISSUE THREAD

**Issue by reykjalin** [wxWidgets/wxWidgets#24056]
### Description

When running an application that's either not in a `.app` bundle, or running the executable directly from within the bundle, the application is not brought to the foreground properly. It's brought to the foreground but does not gain focus. See the videos below:

| Open the bundle | Run the executable directly |
|:---:|:---:|
| <video src="https://github.com/wxWidgets/wxWidgets/assets/13835680/707fa72e-cb00-4fcf-bf8a-9301546e4130"> | <video src="https://github.com/wxWidgets/wxWidgets/assets/13835680/7fdcc73f-3e1a-4ec5-8d00-ea6aa05c40a8" > |

I'm pretty sure the regression happened in https://github.com/wxWidgets/wxWidgets/pull/24003. Before I updated Xcode and `xcode-select` the app wasn't being brought to the foreground at all. After running a `git bisect` in the wx repo I found that https://github.com/wxWidgets/wxWidgets/pull/24003 is what caused the issue. Checking out 4ddea5658d250dcfc416ea5d945e2db23a1246ff had the app brought into the foreground and gain focus correctly.

I initially thought this was because my development libraries on macOS might be out of date so I tried updating Xcode and `xcode-select` and then restarted my computer. After the restart the problem changed into what the recording show above; the app is moved to the foreground but doesn't gain focus. Checking out 4ddea5658d250dcfc416ea5d945e2db23a1246ff no longer fixes the issue.

#### Reproduction steps

1. Checkout and build current `master` or 695ca042895468a8ce6c1a84422882adb3c63ee9.
2. Build the minimal sample like so:

```sh
cd wxWdigets
cmake -B osx_build -G Ninja -DwxBUILD_SAMPLES=SOME -DwxBUILD_SHARED=NO
cmake --build osx_build --target minimal
```

3. Try running the minimal sample via `open osx_build/lib/minimal.app`; note that it pops to the foreground and gains focus.
4. Try running the minimal sample via `osx_build/lib/minimal.app/Contents/MacOS/minimal`; note that the app pops into the foreground **but does not gain focus**.

### Platform and version information

- wxWidgets version  you use: `master`
- wxWidgets port  you use: `wxOSX`
- OS  and its version: macOS Sonoma Version 14.0 (23A344)
- Xcode version: `Version 15.0.1 (15A507)`
- `xcode-select --version`: `xcode-select version 2399.`

**Comment by reykjalin** [comment]
Maybe this is because the `[NSApplicationActivateIgnoringOtherApps](https://developer.apple.com/documentation/appkit/nsapplicationactivationoptions/nsapplicationactivateignoringotherapps?language=objc)` activation option is deprecated? Although changing the value to `[NSApplicationActivateAllWindows](https://developer.apple.com/documentation/appkit/nsapplicationactivationoptions/nsapplicationactivateallwindows?language=objc)` doesn't seem to fix the problem and those are [the only 2 documented options](https://developer.apple.com/documentation/appkit/nsapplicationactivationoptions?language=objc).

**Comment by reykjalin** [comment]
I just realized I had the `CMAKE_OSX_DEPLOYMENT_TARGET` set to `10.0` and after updating it to `14.0` it reverted to the behavior I observed previously; the window doesn't come to the foreground at all:


https://github.com/wxWidgets/wxWidgets/assets/13835680/6be3b409-7288-47de-aa7a-52b6df947e9a

And here's a screenshot of the cmake variable configuration:

<img width="1076" alt="image" src="https://github.com/wxWidgets/wxWidgets/assets/13835680/4dbd7809-1719-406f-a7de-50a7b1fbfecd">

**Comment by reykjalin** [comment]
After some more debugging it turns out this happens in the minimal app because the `wxString app` variable actually finds the bundle even though I'm running the executable directly:

https://github.com/wxWidgets/wxWidgets/blob/2d07dd25b4bf120a7c85baa44e09b17e611a2c08/src/osx/cocoa/utils.mm#L96-L113

So the `activate` conditional is never `true`, so it's not activating the application. I have, however, tested this with my own project (https://git.sr.ht/~reykjalin/Fonn, just in case it's useful) and I'm still seeing the same problem, even after updating the OSX build target; the app isn't activated at all. And in that case the `activate` conditional is `true` (verified with a simple `wxLogDebug` call) so the activation function is definitely being called.

**Comment by csomor** [comment]
ok we run into different behaviours of the deprecated method `activateIgnoringOtherApps` ... the change solved one problem but led to another ...

**Comment by reykjalin** [comment]
To be fair, I think "bundling the app" is a _reasonable_ solution, if just a bit annoying as a requirement on macOS. I'm still surprised that the macOS APIs changed like this.

I've tried to fix this myself, but none of the approaches I've tried have worked, e.g. trying to yield control from the currently active application first as per [this guide](https://developer.apple.com/documentation/appkit/nsapplication/passing_control_from_one_app_to_another_with_cooperative_activation?language=objc) from Apple. Although it's very likely I was just doing that incorrectly because this is my first time trying to interact with the Apple APIs in any meaningful way.

**Comment by csomor** [comment]
I can ask the app for the activation policy, only using the new path for the LSUIElement case referred here

```
      if ( activate ) {
          if ( [NSApp activationPolicy] == NSApplicationActivationPolicyAccessory ) {
              [[NSRunningApplication currentApplication] activateWithOptions: NSApplicationActivateIgnoringOtherApps];
          }
          else {
              [NSApp activateIgnoringOtherApps: YES];
          }
      }
```

**Comment by csomor** [comment]
@vadz this should only have an influence on the activation, but apart from that gdb should work as before

**Comment by reykjalin** [comment]
I can confirm #24062 fixes the issue in my personal project - which hasn't been bundled yet - but it doesn't fix running the executable from within an app that's already in an application bundle, such as with the minimal example that I used in the original issue report.

I'm not so sure we need to fix this for running an executable from within a bundle? I don't know of a use case for running the app that way. I think I will at least consider the issue I'm running into fixed by #24062 👍 

Thank you all for taking a look at this so quickly!

**Comment by vadz** [comment]
> @vadz this should only have an influence on the activation, but apart from that gdb should work as before

Yes, but typically you do want to activate the application before starting to debug it. Anyhow, #24062 fixes this problem for me, thanks!
